src/lib/mina/actions/offchain-state.ts

```diff
@@ -190,271 +211,342 @@ function OffchainState<
 
-        // push action on account update
-        let update = getContract().self;
-        update.body.actions = Actions.pushEvent(update.body.actions, action);
+        let result = await rollup.prove(merkleMap, actions);
+
+        // update internal merkle maps as well
+        // TODO make this not insanely recompute everything
+        // - take new tree from `result`
+        // - update value map in `prove()`, or separately based on `actions`
+        let { merkleMap: newMerkleMap, valueMap: newValueMap } =
+          await fetchMerkleMap(height, internal.contract);
+        internal.merkleMap = newMerkleMap;
+        internal.valueMap = newValueMap;
+
+        return result.proof;
+      },
+
+      async settle(
+        proof: Proof<OffchainStateCommitments, OffchainStateCommitments>
+      ) {
+        // verify the proof
+        proof.verify();
+
+        // check that proof moves state forward from the one currently stored
+        let state =
+          getContract().offchainStateCommitments.getAndRequireEquals();
+        Provable.assertEqual(
+          OffchainStateCommitments,
+          state,
+          proof.publicInput
+        );
+
+        // require that proof uses the correct pending actions
+        getContract().account.actionState.requireEquals(
+          proof.publicOutput.actionState
+        );
+
+        // update the state
+        getContract().offchainStateCommitments.set(proof.publicOutput);
       },
 
-      async get(key) {
-        let keyHash = toKeyHash(prefix, keyType, key);
-        return await get(keyHash, valueType);
+      commitments() {
+        return getContract().offchainStateCommitments;
       },
+
+      fields: Object.fromEntries(
+        Object.entries(config).map(([key, kind], i) => [
+          key,
+          kind.kind === 'offchain-field'
+            ? field(i, kind.type)
+            : map(i, kind.keyType, kind.valueType),
+        ])
+      ) as any,
     };
   }
 
+  const memoizedInstances = new Map<String, OffchainStateInstance<Config>>();
+
   return {
-    setContractInstance(contract) {
-      internal._contract = contract;
-      internal._contractClass =
-        contract.constructor as OffchainStateContractClass;
-    },
-    setContractClass(contractClass) {
-      internal._contractClass = contractClass;
+    init(contractInstance: OffchainStateContract<Config>) {
+      let key = 'COMPILE_TIME';
+      let contractAddress = contractInstance.address;
+      if (contractAddress.isConstant()) {
+        key = contractAddress.toBase58();
+      } else {
+        Provable.asProver(() => {
+          key = contractAddress.toBase58();
+        });
+      }
+      let instance = memoizedInstances.get(key);
+      if (instance === undefined) {
+        instance = OffchainStateInstance();
+        instance.setContractClass(
+          contractInstance.constructor as OffchainStateContractClass<Config>
+        );
+        memoizedInstances.set(key, instance);
+      }
```

Comment:
That was my first instinct, but it continues to cause provable errors:

```
Error: variables only exist inside checked computations
This shouldn't have happened and indicates an internal bug.
    at Bug (file:///Users/coby/Projects/o1js/dist/node/lib/util/errors.js:226:12)
    at assert (file:///Users/coby/Projects/o1js/dist/node/lib/util/errors.js:233:15)
    at toConstantField (file:///Users/coby/Projects/o1js/dist/node/lib/provable/field.js:1083:5)
    at toConstant (file:///Users/coby/Projects/o1js/dist/node/lib/provable/field.js:1076:12)
    at Field.toConstant (file:///Users/coby/Projects/o1js/dist/node/lib/provable/field.js:118:16)
    at Object.fromPublicKey (file:///Users/coby/Projects/o1js/dist/node/lib/ml/conversion.js:64:24)
    at Object.getAccount (file:///Users/coby/Projects/o1js/dist/node/lib/mina/local-blockchain.js:77:52)
    at Module.getAccount (file:///Users/coby/Projects/o1js/dist/node/lib/mina/mina-instance.js:43:27)
    at Object.fetch (file:///Users/coby/Projects/o1js/dist/node/lib/mina/state.js:223:32)
    at onchainActionState (file:///Users/coby/Projects/o1js/dist/node/lib/mina/actions/offchain-state.js:77:81)
    at merkleMaps (file:///Users/coby/Projects/o1js/dist/node/lib/mina/actions/offchain-state.js:86:37)
    at Object.createSettlementProof (file:///Users/coby/Projects/o1js/dist/node/lib/mina/actions/offchain-state.js:222:43)
    at file:///Users/coby/Projects/o1js/dist/node/lib/mina/actions/offchain-contract.unit-test.js:114:50
    at runInstruction (file:///Users/coby/Projects/o1js/dist/node/lib/mina/test/test-contract.js:101:27)
    at execute (file:///Users/coby/Projects/o1js/dist/node/lib/mina/test/test-contract.js:77:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async testLocal (file:///Users/coby/Projects/o1js/dist/node/lib/mina/test/test-contract.js:93:9)
    at async file:///Users/coby/Projects/o1js/dist/node/lib/mina/actions/offchain-contract.unit-test.js:104:1
```

Reply:
This error throws after compiling and deploying successfully, when I try to call `createAccount` in the first transaction to the zkapp after deploying.

Reply:
I landed on the init method because it's the only place within the zkapp itself that will only be executed when I want it to be executed.

Setting the contract instance outside of the zkapp is still possible, e.g.

```
// already-deployed contract
const zkapp = Contract(deployedAddress);
zkapp.offchainState.setContractInstance(zkapp);
```

Reply:
Can you please export IndexedMerkleMapBase as part of this refactoring? It is needed for the serialization of IndexedMerkleMap of any height. Now, only the function IndexedMerkleMap is exported, and I need to get this type using syntax like 
```typescript
export function serializeIndexedMap(
  map: InstanceType<ReturnType<typeof IndexedMerkleMap>>
)
```
https://github.com/zkcloudworker/zkcloudworker-tests/blob/main/src/indexed-map.ts